File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/entrypoints/detectors/HTTPHandlerDetector.kt
/**
 * KSA-162: HTTP Handler Detector — Detects HTTP route handlers.
 */
package com.codeintel.analyzers.entrypoints.detectors

import com.codeintel.analyzers.entrypoints.*

class HTTPHandlerDetector(private val registry: PatternRegistry) {
    private val routeResolver = RouteResolver()

    fun detectFromSymbols(symbols: List<SymbolInput>, framework: String, source: String): List<EntryPoint> {
        val patterns = registry.getFramework(framework) ?: return emptyList()
        val controllerPrefix = findControllerPrefix(symbols, patterns)
        return symbols.mapNotNull { detectHandler(it, framework, patterns, controllerPrefix, source) }
    }

    private fun detectHandler(
        sym: SymbolInput, framework: String, patterns: FrameworkPatterns,
        controllerPrefix: String?, source: String,
    ): EntryPoint? {
        val decorators = sym.decorators ?: emptyList()
        // Decorator-based (NestJS, Spring, FastAPI)
        patterns.decorators?.let { dec ->
            for (handlerPattern in dec.handler) {
                val match = decorators.find { it.contains(handlerPattern) }
                if (match != null) {
                    val (method, path) = extractMethodAndPath(match, handlerPattern)
                    val fullRoute = routeResolver.resolve(controllerPrefix, path)
                    val hasAuth = hasAuthIndicator(decorators, patterns)
                    return EntryPoint(
                        symbolId = sym.id, symbolName = sym.name,
                        filePath = sym.filePath, startLine = sym.startLine,
                        entryType = EntryType.HTTP_HANDLER, framework = framework,
                        httpMethod = method, routePath = path,
                        fullRoute = routeResolver.normalizeParams(fullRoute),
                        middleware = detectMiddleware(decorators),
                        hasAuth = hasAuth, controller = sym.parentName,
                        confidence = Confidence.High,
                    )
                }
            }
        }
        // Call-pattern-based (Express, Ktor, Gin)
        patterns.callPatterns?.let { cp ->
            val ctx = getSymbolContext(source, sym.startLine)
            for (handlerPattern in cp.handler) {
                if (ctx.contains(handlerPattern)) {
                    val method = extractMethodFromCall(handlerPattern)
                    val path = extractPathFromContext(ctx, handlerPattern)
                    val fullRoute = routeResolver.resolve(controllerPrefix, path)
                    return EntryPoint(
                        symbolId = sym.id, symbolName = sym.name,
                        filePath = sym.filePath, startLine = sym.startLine,
                        entryType = EntryType.HTTP_HANDLER, framework = framework,
                        httpMethod = method, routePath = path,
                        fullRoute = routeResolver.normalizeParams(fullRoute),
                        hasAuth = false, confidence = Confidence.Medium,
                    )
                }
            }
        }
        return null
    }

    private fun findControllerPrefix(symbols: List<SymbolInput>, patterns: FrameworkPatterns): String? {
        val prefixes = patterns.decorators?.prefix ?: return null
        for (prefix in prefixes) {
            for (sym in symbols) {
                val match = sym.decorators?.find { it.contains(prefix) }
                if (match != null) return extractPathArg(match)
            }
        }
        return null
    }

    private fun extractMethodAndPath(decorator: String, pattern: String): Pair<String, String> {
        val lower = pattern.lowercase()
        val method = when {
            "post" in lower -> "POST"; "put" in lower -> "PUT"
            "delete" in lower -> "DELETE"; "patch" in lower -> "PATCH"
            else -> "GET"
        }
        return method to extractPathArg(decorator)
    }

    private fun extractPathArg(text: String): String {
        val match = Regex("['\"`]([^'\"`]*)['\"`]").find(text)
        return if (match != null) routeResolver.extractPathFromArg(match.groupValues[1]) else "/"
    }

    private fun extractMethodFromCall(pattern: String): String {
        val p = pattern.lowercase()
        return when {
            "post" in p -> "POST"; "put" in p -> "PUT"
            "delete" in p -> "DELETE"; "patch" in p -> "PATCH"
            else -> "GET"
        }
    }

    private fun extractPathFromContext(context: String, pattern: String): String {
        val idx = context.indexOf(pattern)
        if (idx == -1) return "/"
        val after = context.substring(idx + pattern.length)
        val match = Regex("['\"`]([^'\"`]*)['\"`]").find(after)
        return if (match != null) routeResolver.extractPathFromArg(match.groupValues[1]) else "/"
    }

    private fun getSymbolContext(source: String, startLine: Int): String {
        val lines = source.split("\n")
        val start = maxOf(0, startLine - 3)
        val end = minOf(lines.size, startLine + 5)
        return lines.subList(start, end).joinToString("\n")
    }

    private fun detectMiddleware(decorators: List<String>): List<String> =
        decorators.filter { "UseGuards" in it || "Middleware" in it }

    private fun hasAuthIndicator(decorators: List<String>, patterns: FrameworkPatterns): Boolean =
        patterns.authIndicators.any { indicator -> decorators.any { it.contains(indicator) } }
}

File: backend/src/engine/analyzers/entry-points/detectors/HTTPHandlerDetector.ts
/**
 * KSA-162: HTTP Handler Detector — Detects HTTP route handlers.
 */

import type { EntryPoint, FrameworkPatterns } from '../types.js';
import { PatternRegistry } from '../PatternRegistry.js';
import { RouteResolver } from '../RouteResolver.js';

interface SymbolInput {
  id: number;
  name: string;
  decorators?: string[];
  parentName?: string | null;
  filePath: string;
  startLine: number;
}

export class HTTPHandlerDetector {
  private registry: PatternRegistry;
  private routeResolver: RouteResolver;

  constructor(registry: PatternRegistry) {
    this.registry = registry;
    this.routeResolver = new RouteResolver();
  }

  /** Detect HTTP handlers from symbols and source. */
  detectFromSymbols(symbols: SymbolInput[], framework: string, source: string): EntryPoint[] {
    const patterns = this.registry.getFramework(framework);
    if (!patterns) return [];

    const results: EntryPoint[] = [];
    const controllerPrefix = this.findControllerPrefix(symbols, patterns);

    for (const sym of symbols) {
      const detected = this.detectHandler(sym, framework, patterns, controllerPrefix, source);
      if (detected) results.push(detected);
    }
    return results;
  }

  private detectHandler(
    sym: SymbolInput, framework: string, patterns: FrameworkPatterns,
    controllerPrefix: string | null, source: string
  ): EntryPoint | null {
    const decorators = sym.decorators ?? [];

    // Decorator-based detection (NestJS, Spring, FastAPI)
    if (patterns.decorators) {
      for (const handlerPattern of patterns.decorators.handler) {
        const matchingDec = decorators.find(d => d.includes(handlerPattern));
        if (matchingDec) {
          const { method, path } = this.extractMethodAndPath(matchingDec, handlerPattern);
          const fullRoute = this.routeResolver.resolve(controllerPrefix, path);
          const hasAuth = this.hasAuthIndicator(decorators, patterns);

          return {
            symbol_id: sym.id, symbol_name: sym.name,
            file_path: sym.filePath, start_line: sym.startLine,
            entry_type: 'HTTP_HANDLER', framework,
            http_method: method, route_path: path,
            full_route: this.routeResolver.normalizeParams(fullRoute),
            middleware: this.detectMiddleware(decorators),
            has_auth: hasAuth, controller: sym.parentName ?? null,
            event_name: null, confidence: 'High',
          };
        }
      }
    }

    // Call-pattern-based detection (Express, Ktor, Gin)
    if (patterns.call_patterns) {
      const symContext = this.getSymbolContext(source, sym.startLine);
      for (const handlerPattern of patterns.call_patterns.handler) {
        if (symContext.includes(handlerPattern)) {
          const method = this.extractMethodFromCall(handlerPattern);
          const path = this.extractPathFromContext(symContext, handlerPattern);
          const fullRoute = this.routeResolver.resolve(controllerPrefix, path);

          return {
            symbol_id: sym.id, symbol_name: sym.name,
            file_path: sym.filePath, start_line: sym.startLine,
            entry_type: 'HTTP_HANDLER', framework,
            http_method: method, route_path: path,
            full_route: this.routeResolver.normalizeParams(fullRoute),
            middleware: [], has_auth: false, controller: null,
            event_name: null, confidence: 'Medium',
          };
        }
      }
    }

    return null;
  }

  private findControllerPrefix(symbols: SymbolInput[], patterns: FrameworkPatterns): string | null {
    if (!patterns.decorators?.prefix) return null;
    for (const prefixPattern of patterns.decorators.prefix) {
      for (const sym of symbols) {
        const match = (sym.decorators ?? []).find(d => d.includes(prefixPattern));
        if (match) return this.extractPathArg(match);
      }
    }
    return null;
  }

  private extractMethodAndPath(decorator: string, pattern: string): { method: string; path: string } {
    const lower = pattern.toLowerCase();
    let method = 'GET';
    if (lower.includes('post')) method = 'POST';
    else if (lower.includes('put')) method = 'PUT';
    else if (lower.includes('delete')) method = 'DELETE';
    else if (lower.includes('patch')) method = 'PATCH';
    const path = this.extractPathArg(decorator);
    return { method, path };
  }

  private extractPathArg(text: string): string {
    const match = text.match(/['"`]([^'"`]*)['"`]/);
    if (match) return this.routeResolver.extractPathFromArg(match[1]);
    return '/';
  }

  private extractMethodFromCall(pattern: string): string {
    const p = pattern.toLowerCase();
    if (p.includes('post')) return 'POST';
    if (p.includes('put')) return 'PUT';
    if (p.includes('delete')) return 'DELETE';
    if (p.includes('patch')) return 'PATCH';
    return 'GET';
  }

  private extractPathFromContext(context: string, pattern: string): string {
    const idx = context.indexOf(pattern);
    if (idx === -1) return '/';
    const after = context.slice(idx + pattern.length);
    const match = after.match(/['"`]([^'"`]*)['"`]/);
    return match ? this.routeResolver.extractPathFromArg(match[1]) : '/';
  }

  private getSymbolContext(source: string, startLine: number): string {
    const lines = source.split('\n');
    const start = Math.max(0, startLine - 3);
    const end = Math.min(lines.length, startLine + 5);
    return lines.slice(start, end).join('\n');
  }

  private detectMiddleware(decorators: string[]): string[] {
    return decorators.filter(d => d.includes('UseGuards') || d.includes('Middleware'));
  }

  private hasAuthIndicator(decorators: string[], patterns: FrameworkPatterns): boolean {
    for (const indicator of patterns.auth_indicators) {
      if (decorators.some(d => d.includes(indicator))) return true;
    }
    return false;
  }
}
